### PR TITLE
[quant][embedding qat] eager mode QAT for Embeddings

### LINKED ISSUE #61865: [quant] Support QAT for Embedding/EmbeddingBag
Currently we only support PTQ for embedding modules https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/embedding_ops.py

This issue tracks the support to 
1. add QAT modules for Embedding/EmbeddingBag modules
2. Do the module swap in eager mode to use these modules
3. Update FX quantization flow to support the new modules

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo

### PR DESCRIPTION
**Pull request by b-koopman**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68390
* __->__ #66429

Summary:

Add QAT for Embeddings via eager mode API.

To avoid backward compatibility issues, this has not yet been added to the default QAT mappings.
For examples of use, see tests in "Test Plan" below.

Test Plan:

`pytest test/quantization/core/test_workflow_module.py  -v -k "test_embedding_qat_config"`
`pytest test/quantization/eager/test_quantize_eager_qat.py  -v -k "test_embedding_qat_qconfig_equal" `

Reviewers:
supriyar, HDCharles

Add QAT Embedding to module tests

Add Linear test, update static mapping

remove unneeded test data

fix AO naming

Move test data

Move module mappings out of default

Differential Revision: [D31618284](https://our.internmc.facebook.com/intern/diff/D31618284)

### CODE REVIEW

**Comment by b-koopman** [comment]
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

**Comment by b-koopman** [comment]
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

File: test/quantization/eager/test_quantize_eager_qat.py [hunk 1]
```diff
@@ -109,17 +119,43 @@ def checkQuantized(model):
...
+    def test_defused_embedding_bag_linear(self):
+        for qengine in supported_qengines:
+            with override_quantized_engine(qengine):
+                model = DeFusedEmbeddingBagLinear().train()
+                model = prepare_qat(model)
+                self.checkObservers(model)
+
+                test_only_train_fn(model, self.embed_linear_data_train)
+                # make activation_post_process is not inserted for Embedding
+                self.assertFalse(hasattr(model, "activation_post_process"))
```
Comment (supriyar) [inline test/quantization/eager/test_quantize_eager_qat.py:131]: Shouldn't an activation_post_process be inserted for linear layers here?

File: torch/ao/quantization/quantization_mappings.py [hunk 2]
```diff
@@ -44,6 +44,7 @@
     nn.ELU: nnq.ELU,
     nn.Embedding: nnq.Embedding,
     nn.EmbeddingBag: nnq.EmbeddingBag,
+    nn.Embedding: nnq.Embedding,
```
Comment (supriyar) [inline torch/ao/quantization/quantization_mappings.py:47]: This is the same as line 45

File: torch/nn/qat/modules/embedding_ops.py [hunk 3]
```diff
@@ -3,6 +3,72 @@
...
+
+        return qat_embedding_bag
+
+    def to_float(self):
+        embedding_bag = torch.nn.Embedding(self.num_embeddings, self.embedding_dim, self.padding_idx,
+                                           self.max_norm, self.norm_type, self.scale_grad_by_freq,
+                                           self.sparse, None, self.device, self.dtype)
+        embedding_bag.weight = torch.nn.Parameter(self.weight.detach())
+        embedding_bag.train(self.training)
+        return embedding_bag
```
Comment (supriyar) [inline torch/nn/qat/modules/embedding_ops.py:71]: nit: add a new line after this

File: torch/nn/qat/modules/embedding_ops.py [hunk 4]
```diff
@@ -61,7 +127,6 @@ def from_float(cls, mod):
         qat_embedding_bag = cls(mod.num_embeddings, mod.embedding_dim, mod.max_norm, mod.norm_type,
                                 mod.scale_grad_by_freq, mod.mode, mod.sparse, mod.weight,
                                 mod.include_last_offset, mod.padding_idx, qconfig=qconfig)
-        qat_embedding_bag.weight = mod.weight
```
Comment (supriyar) [inline torch/nn/qat/modules/embedding_ops.py:64]: why did we remove this line?

Comment (b-koopman) [reply torch/nn/qat/modules/embedding_ops.py:64]: This is not required as mod.weight is being passed as a parameter to the constructor.

**Comment by b-koopman** [comment]
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

**Review by HDCharles — APPROVED** [review]

**Comment by b-koopman** [comment]
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

**Comment by b-koopman** [comment]
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

**Comment by b-koopman** [comment]
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

**Comment by b-koopman** [comment]
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

**Comment by b-koopman** [comment]
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

**Review by supriyar — APPROVED** [review]

**Comment by b-koopman** [comment]
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31618284).

**Pull request closed without merging**